File: backend/src/routes/shuttles.js
import express from 'express';
import { pool } from '../db/connection.js';

const router = express.Router();

// Get all shuttles
router.get('/', async (req, res) => {
  try {
    const result = await pool.query(`
      SELECT s.*, c.name as company_name 
      FROM shuttles s
      JOIN companies c ON s.company_id = c.id
      ORDER BY s.name
    `);
    res.json(result.rows);
  } catch (error) {
    console.error('Error fetching shuttles:', error);
    res.status(500).json({ error: 'Failed to fetch shuttles' });
  }
});

// Get shuttle by ID
router.get('/:id', async (req, res) => {
  try {
    const result = await pool.query(`
      SELECT s.*, c.name as company_name 
      FROM shuttles s
      JOIN companies c ON s.company_id = c.id
      WHERE s.id = $1
    `, [req.params.id]);
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Shuttle not found' });
    }
    
    res.json(result.rows[0]);
  } catch (error) {
    console.error('Error fetching shuttle:', error);
    res.status(500).json({ error: 'Failed to fetch shuttle' });
  }
});

// Create new shuttle
router.post('/', async (req, res) => {
  try {
    const { name, company_id, capacity, status } = req.body;
    
    const result = await pool.query(
      'INSERT INTO shuttles (name, company_id, capacity, status) VALUES ($1, $2, $3, $4) RETURNING *',
      [name, company_id, capacity || 50, status || 'active']
    );
    
    res.status(201).json(result.rows[0]);
  } catch (error) {
    console.error('Error creating shuttle:', error);
    res.status(500).json({ error: 'Failed to create shuttle' });
  }
});

// Update shuttle
router.put('/:id', async (req, res) => {
  try {
    const { name, capacity, status } = req.body;
    
    const result = await pool.query(
      `UPDATE shuttles 
       SET name = COALESCE($1, name),
           capacity = COALESCE($2, capacity),
           status = COALESCE($3, status)
       WHERE id = $4
       RETURNING *`,
      [name, capacity, status, req.params.id]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Shuttle not found' });
    }
    
    res.json(result.rows[0]);
  } catch (error) {
    console.error('Error updating shuttle:', error);
    res.status(500).json({ error: 'Failed to update shuttle' });
  }
});

// Delete shuttle
router.delete('/:id', async (req, res) => {
  try {
    const result = await pool.query(
      'DELETE FROM shuttles WHERE id = $1 RETURNING id',
      [req.params.id]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Shuttle not found' });
    }
    
    res.json({ message: 'Shuttle deleted successfully' });
  } catch (error) {
    console.error('Error deleting shuttle:', error);
    res.status(500).json({ error: 'Failed to delete shuttle' });
  }
});

export default router;

File: backend/app/routers/shuttles.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, delete
from sqlalchemy.orm import selectinload, joinedload
from typing import List
from uuid import UUID

from ..database import get_database_session
from ..models import Shuttle, Company
from ..logging_manager import logger_manager
from ..schemas import Shuttle as ShuttleSchema, ShuttleCreate, ShuttleUpdate, MessageResponse
from ..auth import get_current_active_user, AdminUser

router = APIRouter()

@router.get("/public", response_model=List[ShuttleSchema])
async def get_shuttles_public(
    db: AsyncSession = Depends(get_database_session)
):    
    result = await db.execute(
        select(Shuttle, Company.name.label('company_name'))
        .join(Company)
        .order_by(Shuttle.name)
    )
    
    shuttles = []
    for shuttle, company_name in result.all():
        shuttle_dict = {
            "id": shuttle.id,
            "name": shuttle.name,
            "company_id": shuttle.company_id,
            "capacity": shuttle.capacity,
            "status": shuttle.status,
            "created_at": shuttle.created_at,
            "updated_at": shuttle.updated_at,
            "company_name": company_name
        }
        shuttles.append(ShuttleSchema(**shuttle_dict))
    
    return shuttles

@router.get("/", response_model=List[ShuttleSchema])
async def get_shuttles(
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    logger_manager.info("Fetching shuttles list", {"user_id": str(current_user.id)})
    
    result = await db.execute(
        select(Shuttle, Company.name.label('company_name'))
        .join(Company)
        .order_by(Shuttle.name)
    )
    
    shuttles = []
    for shuttle, company_name in result.all():
        shuttle_dict = {
            "id": shuttle.id,
            "name": shuttle.name,
            "company_id": shuttle.company_id,
            "capacity": shuttle.capacity,
            "status": shuttle.status,
            "created_at": shuttle.created_at,
            "updated_at": shuttle.updated_at,
            "company_name": company_name
        }
        shuttles.append(ShuttleSchema(**shuttle_dict))
    
    logger_manager.info("Shuttles retrieved", {"count": len(shuttles)})
    return shuttles

@router.get("/{shuttle_id}", response_model=ShuttleSchema)
async def get_shuttle(
    shuttle_id: UUID,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    result = await db.execute(
        select(Shuttle, Company.name.label('company_name'))
        .join(Company)
        .where(Shuttle.id == shuttle_id)
    )
    row = result.first()
    
    if not row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Shuttle not found"
        )
    
    shuttle, company_name = row
    shuttle_dict = {
        "id": shuttle.id,
        "name": shuttle.name,
        "company_id": shuttle.company_id,
        "capacity": shuttle.capacity,
        "status": shuttle.status,
        "created_at": shuttle.created_at,
        "updated_at": shuttle.updated_at,
        "company_name": company_name
    }
    
    return ShuttleSchema(**shuttle_dict)

@router.post("/", response_model=ShuttleSchema)
async def create_shuttle(
    shuttle_data: ShuttleCreate,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    # Verify company exists
    result = await db.execute(
        select(Company).where(Company.id == shuttle_data.company_id)
    )
    company = result.scalar_one_or_none()
    
    if not company:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Company not found"
        )
    
    new_shuttle = Shuttle(
        name=shuttle_data.name,
        company_id=shuttle_data.company_id,
        capacity=shuttle_data.capacity,
        status=shuttle_data.status
    )
    
    db.add(new_shuttle)
    await db.commit()
    await db.refresh(new_shuttle)
    
    # Return with company name
    shuttle_dict = {
        "id": new_shuttle.id,
        "name": new_shuttle.name,
        "company_id": new_shuttle.company_id,
        "capacity": new_shuttle.capacity,
        "status": new_shuttle.status,
        "created_at": new_shuttle.created_at,
        "updated_at": new_shuttle.updated_at,
        "company_name": company.name
    }
    
    return ShuttleSchema(**shuttle_dict)

@router.put("/{shuttle_id}", response_model=ShuttleSchema)
async def update_shuttle(
    shuttle_id: UUID,
    shuttle_data: ShuttleUpdate,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    # Check if shuttle exists
    result = await db.execute(
        select(Shuttle).where(Shuttle.id == shuttle_id)
    )
    shuttle = result.scalar_one_or_none()
    
    if not shuttle:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Shuttle not found"
        )
    
    # Update only provided fields
    update_data = shuttle_data.dict(exclude_unset=True)
    if update_data:
        await db.execute(
            update(Shuttle)
            .where(Shuttle.id == shuttle_id)
            .values(**update_data)
        )
        await db.commit()
        await db.refresh(shuttle)
    
    # Get company name for response
    company_result = await db.execute(
        select(Company.name).where(Company.id == shuttle.company_id)
    )
    company_name = company_result.scalar()
    
    shuttle_dict = {
        "id": shuttle.id,
        "name": shuttle.name,
        "company_id": shuttle.company_id,
        "capacity": shuttle.capacity,
        "status": shuttle.status,
        "created_at": shuttle.created_at,
        "updated_at": shuttle.updated_at,
        "company_name": company_name
    }
    
    return ShuttleSchema(**shuttle_dict)

@router.delete("/{shuttle_id}", response_model=MessageResponse)
async def delete_shuttle(
    shuttle_id: UUID,
    db: AsyncSession = Depends(get_database_session),
    current_user: AdminUser = Depends(get_current_active_user)
):
    # Check if shuttle exists
    result = await db.execute(
        select(Shuttle).where(Shuttle.id == shuttle_id)
    )
    shuttle = result.scalar_one_or_none()
    
    if not shuttle:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Shuttle not found"
        )
    
    # Delete shuttle
    await db.execute(
        delete(Shuttle).where(Shuttle.id == shuttle_id)
    )
    await db.commit()
    
    return MessageResponse(message="Shuttle deleted successfully")